tuning/trainercontroller/callback.py

```diff
@@ -208,23 +231,26 @@ def _compute_metrics(self, event_name: str, **kwargs):
                 self.metrics[m.get_name()] = m.compute(event_name=event_name, **kwargs)
 
     def _take_control_actions(self, event_name: str, **kwargs):
-        """Invokes the act() method for all the operations registered for a given event. \
-            Note here that the eval() is invoked with `__builtins__` set to None. \
-            This is a precaution to restric the scope of eval(), to only the \
-            fields produced by the metrics.
+        """Invokes the act() method for all the operations registered for a given event.
 
         Args:
             event_name: str. Event name.
             kwargs: List of arguments (key, value)-pairs.
         """
         if event_name in self.control_actions_on_event:
+            evaluator = get_evaluator(metrics=self.metrics)
             for control_action in self.control_actions_on_event[event_name]:
                 rule_succeeded = False
                 try:
-                    # pylint: disable=eval-used
-                    rule_succeeded = eval(
-                        control_action.rule, {"__builtins__": None}, self.metrics
+                    rule_succeeded = evaluator.eval(
+                        expr=control_action.rule_str,
+                        previously_parsed=control_action.rule,
```

Comment:
Can you give an example of where the previously parsed contents would actually be used here? More specifically, if `get_evaluator()` is building a new instance of `EvalWithCompoundTypes` on line 241 - It seems like it is in case there are duplicated rule strings across different control actions for the same event. Is that correct, and are there scenarios where that would be expected?

Reply:
I am not sure I entirely understand the question.

When the `previously_parsed` argument is passed in, the `expr` argument is not used. https://github.com/danthedeckie/simpleeval/blob/2a12b5856d6f70b78dc1ac38840c80c1be6c6c4e/simpleeval.py#L429

We are only giving `expr=control_action.rule_str,` for nice error messages.

Reply:
Thanks that helps - I was partially asking about deduplication, but this
> We are only giving expr=control_action.rule_str, for nice error messages. We can make it an empty string if necessary.

clarifies my confusion. I just didn't understand why both the expr / previously parsed were being passed here when the non-singleton evaluator is being created right before this loop is entered.